controller/cmd/service-mirror/cluster_watcher_test.go

```diff
@@ -0,0 +1,960 @@
+						Name:      "existing-gateway",
+						Namespace: "existing-namespace",
+					},
+				},
+			},
+			expectedLocalServices: []*corev1.Service{
+				mirroredService("service-one-remote", "ns1", "existing-gateway", "existing-namespace", "111", "",
+					[]corev1.ServicePort{
+						{
+							Name:     "port1",
+							Protocol: "TCP",
+							Port:     555,
+						},
+						{
+							Name:     "port2",
+							Protocol: "TCP",
+							Port:     666,
+						},
+					}),
+			},
+			expectedLocalEndpoints: []*corev1.Endpoints{
+				endpoints("service-one-remote", "ns1", "existing-gateway", "existing-namespace", "", nil),
+			},
+			remoteResources: []string{
+				gatewayAsYaml("existing-gateway", "existing-namespace", "222", "192.0.2.127", "incoming-port-wrong", 888, t),
+			},
+		},
+		{
+			testDescription: "create service and endpoints when gateway can be resolved",
+			events: []interface{}{
+				&RemoteServiceCreated{
+
+					service: remoteService("service-one", "ns1", "existing-gateway", "existing-namespace", "111", []corev1.ServicePort{
+						{
+							Name:     "port1",
+							Protocol: "TCP",
+							Port:     555,
+						},
+						{
+							Name:     "port2",
+							Protocol: "TCP",
+							Port:     666,
+						},
+					}),
+					gatewayData: &gatewayMetadata{
+						Name:      "existing-gateway",
+						Namespace: "existing-namespace",
+					},
+				},
+			},
+			expectedLocalServices: []*corev1.Service{
+				mirroredService(
+					"service-one-remote",
+					"ns1",
+					"existing-gateway",
+					"existing-namespace",
+					"111",
+					"222",
+					[]corev1.ServicePort{
+						{
+							Name:     "port1",
+							Protocol: "TCP",
+							Port:     555,
+						},
+						{
+							Name:     "port2",
+							Protocol: "TCP",
+							Port:     666,
+						},
+					}),
+			},
+			expectedLocalEndpoints: []*corev1.Endpoints{
+				endpoints("service-one-remote", "ns1", "existing-gateway", "existing-namespace", "192.0.2.127", []corev1.EndpointPort{
+					{
+						Name:     "port1",
+						Port:     888,
+						Protocol: "TCP",
+					},
+					{
+						Name:     "port2",
+						Port:     888,
+						Protocol: "TCP",
+					},
+				}),
+			},
+			remoteResources: []string{
+				gatewayAsYaml("existing-gateway", "existing-namespace", "222", "192.0.2.127", "incoming-port", 888, t),
+			},
+		},
+	} {
+		tc := tt // pin
+		runTestCase(&tc, t)
+	}
+}
+
+func TestRemoteServiceDeleted(t *testing.T) {
+	for _, tt := range []testCase{
+		{
+			testDescription: "deletes locally mirrored service",
+			events: []interface{}{
+				&RemoteServiceDeleted{
+					Name:      "test-service-remote-to-delete",
+					Namespace: "test-namespace-to-delete",
+				},
+			},
+
+			localResources: []string{
+				mirroredServiceAsYaml("test-service-remote-to-delete", "test-namespace-to-delete", "", "", "", "", nil, t),
+				endpointsAsYaml("test-service-remote-to-delete", "test-namespace-to-delete", "", "", "", nil, t),
+			},
+		},
```

Comment:
Given `expectedLocalServices` and `expectedLocalEndpoints` is empty in this test, the loops in `runTestCase()` won't run and there is no check performed verifying if that service and endpoint are really gone. 
One way of verifying could be checking if those expected slices are empty, and if so, loop against `localResources` and verify the local API can't retrieve any service/endpoint from that slice...

Reply:
Thanks for catching that @alpeb !